File: 02/code/objc/Lection02/Lection02/Greeting.m
//
//  Greeting.m
//  Lection02
//
//  Created by d.taraev on 16.03.17.
//  Copyright © 2017 mail.ru. All rights reserved.
//

#import "Greeting.h"

@implementation Greeting

- (instancetype)initWithName:(NSString *)name gender:(Gender)gender age:(NSInteger)age {
    
    if (self=[super init]) {
        
        _name = [name copy];
        _gender = gender;
        _age = age;
    }
    return self;
}

- (instancetype)init {
    
    return [self initWithName:@"" gender:unknown age:-1];
}

- (NSString *)greeting {
    
    if (self.age > 0) {
        
        NSString *genderString;
        switch (self.gender) {
            case male:
                genderString = @"милый";
                break;
            case female:
                genderString = @"милая";
                break;
            case unknown:
                genderString = @"мил(ый/ая)";
                break;
                
            default:
                break;
        }
        
        NSString *ageString;
        if (self.age < 20) {
            switch (self.gender) {
                case male:
                    ageString = @"юный";
                    break;
                case female:
                    ageString = @"юная";
                    break;
                case unknown:
                    ageString = @"юн(ый/ая)";
                    break;
                    
                default:
                    break;
            }
        } else
            ageString = @"";
        
        return [NSString stringWithFormat:@"Привет, %@ %@ %@!!!", genderString, ageString, self.name];
        
    } else
        return @"";
}

@end

File: 02/code/Lection02/Greeting.swift
//
//  Greeting.swift
//  Lection02
//
//  Created by Gennady Evstratov on 11/03/2017.
//  Copyright © 2017 Mail. All rights reserved.
//

import Foundation

enum Gender: Int {
    case male = 0
    case female = 1
    case unknown = 2
}

struct Greeting {
    let name: String
    let gender: Gender
    let age: Int
    
    static let empty = Greeting(name: "", gender: .unknown, age: -1)
    
    var greeting: String {
        guard age > 0 else {
            return ""
        }
        
        let genderString: String
        switch gender {
        case .male:
            genderString = "милый"
            
        case .female:
            genderString = "милая"
            
        case .unknown:
            genderString = "мил(ый/ая)"
        }
        
        let ageString: String
        if age < 20 {
            switch gender {
            case .male:
                ageString = "юный"
                
            case .female:
                ageString = "юная"
                
            case .unknown:
                ageString = "юн(ый/ая)"
            }
        } else {
            ageString = ""
        }
        
        return "Привет, \(genderString) \(ageString) \(name)!!!"
    }
}

